subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c

```diff
@@ -416,6 +423,46 @@ int nct_mqtt_connect(void)
 	return mqtt_connect(&nct.client);
 }
 
+static int publish_get_payload(struct mqtt_client *client, size_t length)
+{
+	u8_t *buf = nct.payload_buf;
+	u8_t *end = buf + length;
+	struct pollfd fds;
+
+	if (length > sizeof(nct.payload_buf)) {
+		return -EMSGSIZE;
+	}
+
+	if (client->transport.type == MQTT_TRANSPORT_SECURE) {
+		fds.fd = client->transport.tls.sock;
+	} else {
+		fds.fd = client->transport.tcp.sock;
+	}
+
+	fds.events = POLLIN;
+
+	while (buf < end) {
+		int ret = mqtt_read_publish_payload(client, buf, end - buf);
+		if (ret < 0) {
+			if (ret == -EAGAIN) {
+				LOG_DBG("mqtt_read_publish_payload: EAGAIN");
+				poll(&fds, 1, K_FOREVER);
```

Comment:
Have you confirmed that `poll()` is actually blocking here until there is data on the socket? I've noticed that `poll` returns with error code `0` immediately after being called, indicating it timed out, regardless of what `timeout` is set to. However, setting the timeout to zero, `poll` appearently blocks as if it was set to `K_FOREVER` / `-1`. 

Talking to @gbakke and @metc, it was confirmed that timeouts on `poll` when sockets are offloaded don't behave the way one would expect, but the feature will be implemented in near future.

Reply:
No I have not verified `nrf_poll` behavior on my own. I didn't deeply anylaze `nrf_poll` internals, but I guess some misbehave on it's side might be result of the fact, that some parts of bsdlib os layer are not implemented:
https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/d03c91c4c1e9655b6a4d20853d94daec1dfb16e3/lib/bsdlib/bsd_os.c#L52

Reply:
And it's really hard to spot such issues for people that were not involved into bsdlib development, due to fact that some parts of the API are still not documented:
https://github.com/NordicPlayground/nrfxlib/blob/4173ecf39b5798100a3859dafbb0bca9ea6b24fc/bsdlib/include/bsd_os.h#L23

Reply:
Yes, I saw some unexpected behaviour so just did a quick check to see that the timeout was actually respected. There's a a branch hopefully fixing this. If we can test it and confirm, it can be turned into a PR soon:
https://github.com/gbakke/fw-nrfconnect-nrf/commits/bsd_os_timedwait

Reply:
Briefly tested. and the branch I linked to seems to fix the timeout issues with `poll` using `K_FOREVER`. Will do more testing tomorrow.